### PR TITLE
Add support for custom mapping

### PR DESCRIPTION
**Pull request by lussoluca**
This PR adds two new configurations: ** analysis_file_path** and ** mapping_file_path** to specify the files that contains a custom analysis and mapping json to be used during index creation. It also allows to customize the alias name (with the ** alias_name** configuration).

### CODE REVIEW

**Comment by robsears** [comment]
Thanks @lussoluca, this looks awesome. It's very similar to something we at OMC talked about doing but never got around to. So thanks for the leg work!

If I'm reading this right, the code looks for a `analysis_file_path` or `mapping_file_path` in the site config. If nothing is specified, then it defaults to `./mapping/analysis.json` or `./mapping/mapping.json` respectively. This means that there will always be a file path defined for both mappings and analysis.

This means that `if File.exist?(@@mapping_file_path)` and `if File.exist?(@@analysis_file_path)` will always evaluate to true, unless the user specifically configures the site to point at a file that doesn't exist (or is nil or something).

Consequently, these `*.json` files will become the defaults for all Searchyll users, unless the user specifically overrides them somehow.

I would like to avoid that pattern if possible. Something like this instead:

```ruby
site.config['elasticsearch']['analysis_file_path']

...

if !@@analysis_file_path.nil? && File.exist?(@@analysis_file_path)
```

An update to the README explaining the new features would also be great.

**Comment by ninomarrazzo** [comment]
Hi @robsears, i'm Nino.
Last commits to this PR embeddes what you rightly observed: now if there isn't specified a value (in jekyll's _config.yml file) for mapping_file_path or analysis_file_path (or both two) does simply nothing.

Regards README as soon as possible i will write what's new.

File: lib/searchyll/indexer.rb [hunk 1]
```diff
@@ -30,6 +31,10 @@ def initialize(configuration)
       self.working       = true
       self.timestamp     = Time.now
       self.batch_size    = BATCH_SIZE
+      # Mapping fields JSON file path
+      @@mapping_file_path = configuration.elasticsearch_mapping_fields
+      # Analysis fields JSON file path
+      @@analysis_file_path = configuration.elasticsearch_analysis_fields
```
Comment (drusellers) [inline lib/searchyll/indexer.rb]: What's the intent behind making this a class method if you are going to adjust it at the time of instance construction? Could this not be `self. analysis_file_path` like the rest of the fields?

File: lib/searchyll/mapping/analysis.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (drusellers) [inline lib/searchyll/mapping/analysis.json]: Is this file necessary? or maybe renamed as 'example'?

File: lib/searchyll/mapping/fields.json [hunk 3]
```diff
@@ -0,0 +1,106 @@
+{
```
Comment (drusellers) [inline lib/searchyll/mapping/fields.json]: Is this file necessary? or maybe renamed as 'example'?

File: lib/searchyll/indexer.rb [hunk 4]
```diff
@@ -1,5 +1,6 @@
 require 'json'
 require 'net/http'
+require 'yaml'
```
Comment (drusellers) [inline lib/searchyll/indexer.rb]: Jekyll already uses SafeYAML - should we just use that? `SafeYAML.load(content)`

File: lib/searchyll/indexer.rb [hunk 5]
```diff
@@ -111,20 +116,40 @@ def http_start
...
-          refresh_interval:   -1
+          refresh_interval: -1
         }
-      }.to_json # TODO: index settings
+      }.to_json
+      
+      # Add mapping fields to the index
+      if mapping_fields
+        create_index.class.instance_eval('attr_accessor :mappings')
+        create_index.mappings = mapping_fields
```
Comment (drusellers) [inline lib/searchyll/indexer.rb]: What's the thought behind this method of mutating the class - rather than just changing how the hash is made that is passed to `create_index.body` - This seems odd to me, but maybe I could learn something about ruby. :)

**Comment by allizad** [comment]
@lussoluca thanks for your PR on this — #36, recently merged, solved this similarly.

File: lib/searchyll/indexer.rb (hunk 1 above)
Comment (reinchek) [reply lib/searchyll/indexer.rb]: @drusellers Sorry for the late answer! There's not a specified reason for which i've done so in that way. Maybe the only reason is that, not being a ruby developer, during developing I couldn't get self.analysis_file_path (and the other) to works. But now, in hindsight, i think that the problem wasn't relative to variables' scope... so maybe is the case to use self keyword.

File: lib/searchyll/mapping/analysis.json (hunk 2 above)
Comment (reinchek) [reply lib/searchyll/mapping/analysis.json]: @drusellers No this file isn't strictly necessary... maybe renamed as example is better.

File: lib/searchyll/mapping/fields.json (hunk 3 above)
Comment (reinchek) [reply lib/searchyll/mapping/fields.json]: @drusellers No this file isn't strictly necessary... maybe renamed as example is better.

File: lib/searchyll/indexer.rb (hunk 4 above)
Comment (reinchek) [reply lib/searchyll/indexer.rb]: @drusellers Unluckly i don't know how SafeYAML works but if this doesn't akter the parsed object's structure it is still fine.

File: lib/searchyll/indexer.rb (hunk 5 above)
Comment (reinchek) [reply lib/searchyll/indexer.rb]: @drusellers Same thing as what said in this [comment](https://github.com/omc/searchyll/pull/34#discussion_r251320856): for unknown reasons in that moment i couldn't get it to works differently. I'm sure that yours ruby's knowledge are better then mine so i counting on it to solves this odd solution. Thank you!
